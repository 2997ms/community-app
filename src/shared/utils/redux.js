/**
 * Redux-related helpers.
 */

import _ from 'lodash';

/**
 * Given any Flux Standard Action (FSA) with promise as the payload, it returns
 * a promise which resolves into the FSA result object.
 * @param {Object} action
 * @return Promise which resolves to the operation result.
 */
export function toFSA(action) {
  return action.payload.then(data => ({
    ...action,
    payload: data,
  }), error => ({
    ...action,
    payload: error,
    error: true,
  }));
}

/**
 * Given a mapping between reducer names and their promises, this function
 * waits till resolution of all promises and returns the mapping between
 * reducer names and the resolved reducers.
 * @param {Object} promises Object with promises of reducers.
 * @return Object with reducers.
 */
export function resolveReducers(promises) {
  return Promise.all(_.values(promises)).then((reducers) => {
    const map = {};
    _.keys(promises).forEach((key, index) => {
      map[key] = reducers[index];
    });
    return map;
  });
}

/* We don't any `main` function in this module to export it by default. */
export default undefined;
