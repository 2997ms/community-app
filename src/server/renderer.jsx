/**
 * This module implements ExpressJS middleware for server-side rendering of
 * the App.
 */

import config from 'config';
import React from 'react';
import ReactDOM from 'react-dom/server';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import serializeJs from 'serialize-javascript';
import { isDev } from 'utils/isomorphy';

import App from '../shared';

/* This is always initial state of the store. Later we'll have to provide a way
 * to put the store into correct state depending on the demanded route. */
import storeFactory from '../shared/store-factory';

export default (req, res) => {
  storeFactory(req).then((store) => {
    const context = {};
    const appHtml = ReactDOM.renderToString((
      <Provider store={store}>
        <StaticRouter
          context={context}
          location={req.url}
        >
          <App />
        </StaticRouter>
      </Provider>
    ));
    if (context.status) res.status(context.status);
    res.send((
      `<!DOCTYPE html>
      <html>
        <head>
          <title>Topcoder</title>
          ${
            /* In dev environment styles are embed into JS bundle,
              and we should not include this link into the page, otherwise
              the page will initially load production /style.css and it may
              result in some wierd behavior while loading. */
            isDev() ? '' : '<link rel="stylesheet" href="/style.css" />'
          }
        </head>
        <body>
          <div id="react-view">${appHtml}</div>
          <script type="application/javascript">
            window.CONFIG = ${serializeJs(config, { isJSON: true })}
            window.ISTATE = ${serializeJs(store.getState(), { isJSON: true })}
          </script>
          <script type="application/javascript" src="/bundle.js"></script>
        </body>
      </html>`
    ));
  });
};
