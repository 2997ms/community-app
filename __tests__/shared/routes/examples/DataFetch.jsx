import React from 'react';
import renderer from 'react-test-renderer';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import _ from 'lodash';

import DataFetchRoute from 'routes/examples/DataFetch';

const reducer = _.identity;

const store = createStore(reducer);

let originalFetch;

beforeAll(() => {
  originalFetch = global.fetch;
  global.fetch = jest.fn(() =>
  Promise.resolve({ json: () => [] }));
});

afterAll(() => {
  global.fetch = originalFetch;
});

test('matches snapshots', () => {
  const cmp = renderer.create(
    <Provider store={store}>
      <StaticRouter context={{}}>
        <DataFetchRoute />
      </StaticRouter>
    </Provider>,
  );
  expect(cmp.toJSON()).toMatchSnapshot();
});
