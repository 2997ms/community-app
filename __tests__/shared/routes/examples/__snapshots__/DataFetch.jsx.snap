// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots 1`] = `
<div>
  Initial State: no data, and not loading yet.
</div>
`;
